docs/proposals/templates.md

```diff
@@ -0,0 +1,404 @@
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest release of this document can be found
+[here](http://releases.k8s.io/release-1.1/docs/proposals/templates.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Templates+Parameterization: Repeatedly instantiating user-customized application topologies.
+
+## Motivation
+
+Addresses https://github.com/kubernetes/kubernetes/issues/11492
+
+Kubernetes gives developers a platform on which to run images and many configuration objects to control those images, but 
+constructing a cohesive application made up of images and configuration objects is currently difficult.  Applications 
+require:
+
+* Information sharing between images (eg one image provides a DB service, another consumes it)
+* Configuration/tuning settings (memory sizes, queue limits)
+* Unique/customizable identifiers (service names, routes)
+
+Application authors know which values should be tunable and what information must be shared, but there is currently no 
+consistent way for an application author to define that set of information so that application consumers can easily deploy 
+an application and make appropriate decisions about the tunable parameters the author intended to expose.  
+
+Furthermore, even if an application author provides consumers with a set of API object definitions (eg a set of yaml files) 
+it is difficult to build a UI around those objects that would allow the deployer to modify names in one place without 
+potentially breaking assumed linkages to other pieces.  There is also no prescriptive way to define which configuration 
+values are appropriate for a deployer to tune or what the parameters control.
+
+## Use Cases
+
+### Use cases for templates in general
+
+* Providing a full baked application experience in a single portable object that can be repeatably deployed in different environments.
+  * eg Wordpress deployment with separate database pod/replica controller
+  * Complex service/replication controller/volume topologies
+* Bulk object creation
+* Provide a management mechanism for deleting/uninstalling an entire set of components related to a single deployed application
+* As a developer I want to select the template to use from a list of available templates
+* Enabling the creation of user interfaces that can guide an application deployer through the deployment process with descriptive help about the configuration value decisions they are making, and useful default values where appropriate
+
```

Comment:
Additional use case:
* Capturing the topology of an application from template expansion as metadata within the cluster, so that
   * visualizations can be provided that expose application structure, 
   * developers can inspect a cluster to determine what's running in it,
   * tools can use the metadata to aggregate and analyze signals.